### PR TITLE
Enable DH in generate_psa_tests.py

### LINKED ISSUE #7812: Automatically generate PSA tests also for DH keys
### Suggested enhancement

While working on #7771 I noticed this comment in `generate_psa_tests.py`:

https://github.com/Mbed-TLS/mbedtls/blob/a3a0025e1831b9d22e8587d88a4d4d42dfad38bb/tests/scripts/generate_psa_tests.py#L143

However due to the recent development on DH keys, I'm not sure this comment (and following actions) is still correct. If this is the case, then the following lines should be removed:

```
        constructors.key_types.discard('PSA_KEY_TYPE_DH_KEY_PAIR')
        constructors.key_types.discard('PSA_KEY_TYPE_DH_PUBLIC_KEY')
```

### Justification

Mbed TLS needs this in order to fully cover asymmetric keys testing.

### PR DESCRIPTION
**Pull request by mpg**
## Description

Resolves #7812: add support for DH keys to `generate_psa_tests.py`.

## PR checklist

- [x] **changelog** not required - tests only
- [x] **backport** not required - we only have that feature in development
- [x] **tests** provided - that's what the PR is about :)

### CODE REVIEW

File: tests/scripts/generate_psa_tests.py [hunk 1]
```diff
@@ -265,12 +263,16 @@ def test_cases_for_key_type_not_supported(
...
                      'PSA_KEY_TYPE_ECC_PUBLIC_KEY')
+    DH_KEY_TYPES = ('PSA_KEY_TYPE_DH_KEY_PAIR',
+                    'PSA_KEY_TYPE_DH_PUBLIC_KEY')
 
     def test_cases_for_not_supported(self) -> Iterator[test_case.TestCase]:
         """Generate test cases that exercise the creation of keys of unsupported types."""
         for key_type in sorted(self.constructors.key_types):
             if key_type in self.ECC_KEY_TYPES:
                 continue
+            if key_type in self.DH_KEY_TYPES:
```
Comment (AndrzejKurek) [inline tests/scripts/generate_psa_tests.py:274]: Minor: this could be done in one line with the previous if.

**Review by AndrzejKurek — DISMISSED** [review]
Looks good comparing to the way `ECC_KEY_TYPES` are handled.

File: scripts/mbedtls_dev/crypto_knowledge.py [hunk 2]
```diff
@@ -138,6 +138,9 @@ def is_public(self) -> bool:
         """Whether the key type is for public keys."""
         return self.name.endswith('_PUBLIC_KEY')
```
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/crypto_knowledge.py:140]: Minor: please tweak line to `short_expression` to abbreviate `DH_FAMILY` like `ECC_FAMILY`.

File: tests/scripts/generate_psa_tests.py [hunk 3]
```diff
@@ -265,12 +263,16 @@ def test_cases_for_key_type_not_supported(
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_psa_tests.py:263]: The dependency for DH key pairs is `PSA_WANT_KEY_TYPE_DH_KEY_PAIR`, so the tests are not executed. This needs to change to the new suffixes, like is already the case for ECC.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
A couple of minor issues. Also I haven't ensured that `test_suite_psa_crypto_not_supported.generated.data` looks right because of https://github.com/Mbed-TLS/mbedtls/issues/7915.

**Comment by mpg** [comment]
Just pushed a new version that I think fixes #7915. It won't pass CI yet as `make test` doesn't pass in the default config: `PSA key_agreement FFDH: incompatible with DH_KEY_PAIR(RFC7919)` from `test_suite_psa_crypto_op_fail.generated` is failing and I need to look into this, but I wanted to push the current state in order to share the fix for #7915.

**Comment by gilles-peskine-arm** [comment]
> PSA key_agreement FFDH: incompatible with DH_KEY_PAIR(RFC7919) from test_suite_psa_crypto_op_fail.generated is failing

That means in `crypto_knowledge.py`, in `KeyType`, `can_do` isn't doing its job. And indeed it's missing a case for FFDH.

**Review by gilles-peskine-arm — COMMENTED** [review]
43257747731861e40fc171070a504d61bc6409c0 looks good to me, but now `generate_psa_tests.py` no longer needs the `MBEDTLS_GENPRIME` insertion hack.

**Comment by mpg** [comment]
Indeed, it became unnecessary 15h ago when #7902 was merged. I've rebased to fix the conflict and remove the last commit.

**Review by gilles-peskine-arm — APPROVED** [review]
Looks good to me, except for wrong dependencies in `test_suite_psa_crypto_not_supported.generated.data` which is a pre-existing issue that will be resolved with https://github.com/Mbed-TLS/mbedtls/issues/7915.

**Comment by mpg** [comment]
Well, I did try to resolve #7915 with this PR (as noted in the PR description), specifically [this commit](https://github.com/Mbed-TLS/mbedtls/pull/7909/commits/c154a043bb299847ec6c36b32517f0d99b032a1a) (whose message is out-of-date after rebasing btw). What did I miss?

**Comment by gilles-peskine-arm** [comment]
> Well, I did try to resolve #7915 with this PR (as noted in the PR description), specifically this commit (whose message is out-of-date after rebasing btw). What did I miss?

Oops, I missed that. Well, #7915 is not resolved, but I think the current state is good to merge anyway as long as we don't close the issue.

Observation:
```
# Take config-suite-b.h plus PSA (which is executed by test-ref-configs.pl on the CI)
mbedtls-prepare-build -p debug -d build-psa-suite-b --config-file=configs/config-suite-b.h --config-set=MBEDTLS_PSA_CRYPTO_C,MBEDTLS_USE_PSA_CRYPTO
make -C build-psa-suite-b tests/test_suite_psa_crypto_not_supported.generated.run
```

Comparing the output on development and this PR:
```
diff ../../f1c032adba124fecbb43ff52d3e08f7157ae69e1.txt ../../182eb1514e49a1dc1eafffd83ab7abf03e3c6591.txt
1c1
< make: Entering directory '/home/gilpes01/z/dev/mbedtls/main/build-psa-suite-b'
---
> make: Entering directory '/home/gilpes01/z/dev/mbedtls/reviewing/7909-dh-generate-psa-tests/build-psa-suite-b'
224a225,254
> PSA import DH_KEY_PAIR(RFC7919) 2048-bit type not supported ....... ----
> PSA generate DH_KEY_PAIR(RFC7919) 2048-bit type not supported ..... ----
> PSA import DH_KEY_PAIR(RFC7919) 3072-bit type not supported ....... ----
> PSA generate DH_KEY_PAIR(RFC7919) 3072-bit type not supported ..... ----
> PSA import DH_KEY_PAIR(RFC7919) 4096-bit type not supported ....... ----
> PSA generate DH_KEY_PAIR(RFC7919) 4096-bit type not supported ..... ----
> PSA import DH_KEY_PAIR(RFC7919) 6144-bit type not supported ....... ----
> PSA generate DH_KEY_PAIR(RFC7919) 6144-bit type not supported ..... ----
> PSA import DH_KEY_PAIR(RFC7919) 8192-bit type not supported ....... ----
> PSA generate DH_KEY_PAIR(RFC7919) 8192-bit type not supported ..... ----
> PSA import DH_KEY_PAIR(RFC7919) 2048-bit group not supported ...... ----
> PSA generate DH_KEY_PAIR(RFC7919) 2048-bit group not supported .... ----
> PSA import DH_KEY_PAIR(RFC7919) 3072-bit group not supported ...... ----
> PSA generate DH_KEY_PAIR(RFC7919) 3072-bit group not supported .... ----
> PSA import DH_KEY_PAIR(RFC7919) 4096-bit group not supported ...... ----
> PSA generate DH_KEY_PAIR(RFC7919) 4096-bit group not supported .... ----
> PSA import DH_KEY_PAIR(RFC7919) 6144-bit group not supported ...... ----
> PSA generate DH_KEY_PAIR(RFC7919) 6144-bit group not supported .... ----
> PSA import DH_KEY_PAIR(RFC7919) 8192-bit group not supported ...... ----
> PSA generate DH_KEY_PAIR(RFC7919) 8192-bit group not supported .... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 2048-bit type not supported ..... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 3072-bit type not supported ..... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 4096-bit type not supported ..... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 6144-bit type not supported ..... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 8192-bit type not supported ..... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 2048-bit group not supported .... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 3072-bit group not supported .... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 4096-bit group not supported .... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 6144-bit group not supported .... ----
> PSA import DH_PUBLIC_KEY(RFC7919) 8192-bit group not supported .... ----
228,229c258,259
< PASSED (222 / 222 tests (169 skipped))
< make: Leaving directory '/home/gilpes01/z/dev/mbedtls/main/build-psa-suite-b'
---
> PASSED (252 / 252 tests (199 skipped))
> make: Leaving directory '/home/gilpes01/z/dev/mbedtls/reviewing/7909-dh-generate-psa-tests/build-psa-suite-b'
```

Resolving 7915 would mean that the DH tests pass here, and also the ECC test cases for curves that are not supported in development pass here. But they're still skipped.

**Review by tom-daubney-arm — APPROVED** [review]
LGTM, thanks.

**Comment by gilles-peskine-arm** [comment]
I've removed the statement about #7915 from the issue description since I approved thinking that this would resolve #7815 but not #7915.

**Pull request merged**